File: app/src/main/java/cm/aptoide/pt/editorial/EditorialRepository.java
package cm.aptoide.pt.editorial;

import rx.Single;

/**
 * Created by D01 on 29/08/2018.
 */

public class EditorialRepository {

  private final EditorialService editorialService;
  private EditorialViewModel cachedEditorialViewModel;

  public EditorialRepository(EditorialService editorialService) {
    this.editorialService = editorialService;
  }

  public Single<EditorialViewModel> loadEditorialViewModel(String cardId) {
    if (cachedEditorialViewModel != null) {
      return Single.just(cachedEditorialViewModel);
    }
    return editorialService.loadEditorialViewModel(cardId)
        .map(editorialViewModel -> {
          if (!editorialViewModel.hasError() && !editorialViewModel.isLoading()) {
            cachedEditorialViewModel = editorialViewModel;
          }
          return editorialViewModel;
        });
  }
}

File: app/src/main/java/cm/aptoide/pt/editorial/EditorialRepository.kt
package cm.aptoide.pt.editorial

import rx.Single

class EditorialRepository(private val editorialService: EditorialService) {

  private var cachedEditorialViewModel: EditorialViewModel? = null;

  fun loadEditorialViewModel(editorialLoadSource: EditorialLoadSource): Single<EditorialViewModel> {
    return if (cachedEditorialViewModel != null)
      Single.just(cachedEditorialViewModel)
    else
      when (editorialLoadSource) {
        is CardId -> editorialService.loadEditorialViewModel(
            editorialLoadSource.cardId).doOnError { throwable -> throwable.printStackTrace() }.map { editorialViewModel ->
          saveResponse(editorialViewModel)
        }
        is Slug -> editorialService.loadEditorialViewModelWithSlug(
            editorialLoadSource.slug).doOnError { throwable -> throwable.printStackTrace() }.map { editorialViewModel ->
          saveResponse(editorialViewModel)
        }
      }
  }

  private fun saveResponse(editorialViewModel: EditorialViewModel): EditorialViewModel {
    if (!editorialViewModel.hasError() && !editorialViewModel.isLoading) {
      cachedEditorialViewModel = editorialViewModel
    }
    return editorialViewModel
  }
}